core/modules/replica/MessengerConnector.cc

```diff
@@ -0,0 +1,640 @@
+        boost::bind(
+            &MessengerConnector::responseReceived,
+            shared_from_this(),
+            boost::asio::placeholders::error,
+            boost::asio::placeholders::bytes_transferred
+        )
+    );
+}
+
+void MessengerConnector::responseReceived(boost::system::error_code const& ec,
+                                          size_t bytes_transferred) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "responseReceived"
+         << "  _currentRequest->id=" << (_currentRequest ? _currentRequest->id : "")
+         << " error_code=" << ec);
+
+    // The notification if any should be happening outside the lock guard
+    // to prevent deadlocks
+    //
+    // FIXME: this may be reconsidered because some requests may take
+    // substantial amount of time to process the notification. This may
+    // result in blocking processing other high-priority requests waiting
+    // in the input queue. The simplest approach would be probably launch
+    // the notification in a separate (new) thread.
+
+    WrapperBase::Ptr request2notify;
+    {
+        util::Lock lock(_mtx, context() + "responseReceived");
+
+        // Check if the request was cancelled while still in flight.
+        // If that happens then _currentRequest should already be nullified
+        // and request removed from all data structures.
+
+        if (isAborted(ec)) {
+            restart(lock);
+            return;
+        }
+
+        if (_currentRequest) {
+
+            // The requst is still valid
+            if (ec) {
+
+                // If something bad happened along the line then we should notify a caller about
+                // a problem with the request, remove the request from all data structures and
+                // restart the communication.
+
+                request2notify = _currentRequest;
+
+                _id2request.erase(_currentRequest->id);
+                _currentRequest = nullptr;
+
+                restart(lock);
+
+            } else {
+
+                // Receive response header into the temporary buffer
+
+                if (syncReadVerifyHeader(lock,
+                                         _inBuffer,
+                                         _inBuffer.parseLength(),
+                                         _currentRequest->id)) {
+
+                    // If something bad happened along the line then we should notify a caller about
+                    // a problem with the request, remove the request from all data structures and
+                    // restart the communication.
+
+                    request2notify = _currentRequest;
+
+                    _id2request.erase(_currentRequest->id);
+                    _currentRequest = nullptr;
+
+                    restart(lock);
+
+                } else {
+
+                    size_t bytes;
+                    if (syncReadFrame(lock,
+                                      _inBuffer,
+                                      bytes)) {
```

Comment:
do the vars need to be on separate lines?

Reply:
They don't have to be. However, I've not seen any specific requirements in the Style Guide in this subject.  I often do this to improve readability of my code, even if this adds extra lines :-)